### PR TITLE
(#267) Don’t show “Public X” header without contents

### PR DESCRIPTION
**Pull request by danielparks**
Previously, the “Public X” header would be displayed if there were any private X regardless of whether there were any public X.

For example, having a private function would make both the “Public Functions” and “Private Functions” headers appear, even if there weren’t any public functions.

This changes it so that there are three conditions:

  * **Only public X:** no “Public” or “Private“ headers are displayed; the X are listed with links to their documentation. (Public is implied.)
  * **Only private X:** a “Private X” header is is displayed with a list of X. There are no links to documentation for private APIs.
  * **Both public and private X:** both “Public X” and “Private X” headers are displayed. The public Xs are listed with links to their documentation; the private Xs are just listed with no links.

In other words, if there are no private Xs, then it just lists the public once. Otherwise, it splits them up under public/private headers but avoids showing a header if it’s contents will be empty.

This also radically simplifies and removes a bunch of boilerplate code around rendering these sections.

### CODE REVIEW

File: lib/puppet-strings/markdown/table_of_contents.rb [hunk 1]
```diff
@@ -3,24 +3,17 @@
 module PuppetStrings::Markdown
   module TableOfContents
     def self.render
-      final = "## Table of Contents\n\n"
+      template = File.join(File.dirname(__FILE__), 'templates/table_of_contents.erb')
```
Comment (ekohl) [inline lib/puppet-strings/markdown/table_of_contents.rb:6]:
```suggestion
      template = File.join(__dir__, 'templates', 'table_of_contents.erb')
```

File: lib/puppet-strings/markdown/table_of_contents.rb [hunk 2]
```diff
@@ -3,24 +3,17 @@
...
-      PuppetStrings::Markdown::PuppetTasks,
-      PuppetStrings::Markdown::PuppetPlans].each do |r|
-        toc = r.toc_info
-        group_name = toc.shift
-        group = toc
-        priv = r.contains_private?
-
-        template = File.join(File.dirname(__FILE__), 'templates/table_of_contents.erb')
-        final += PuppetStrings::Markdown.erb(template).result(binding)
+        PuppetStrings::Markdown.erb(template).result(binding)
```
Comment (ekohl) [inline lib/puppet-strings/markdown/table_of_contents.rb:13]: I wonder if you can also store this outside of the loop (just like the `template` variable) and render it multiple times

**Review by ekohl — COMMENTED** [review]
Nice deduplication. Looks like there's room for more. Some minor details inline.

File: lib/puppet-strings/markdown/table_of_contents.rb (hunk 2 above)
Comment (danielparks) [reply lib/puppet-strings/markdown/table_of_contents.rb:13]: Good question!

**Review by ekohl — APPROVED** [review]

**Comment by danielparks** [comment]
I realized that it was a lot clearer to pull `PuppetStrings::Markdown::TableOfContents` into the general renderer, so I did. I also did some future-rubocop-related clean ups, though it’s still going to take some work to rebase this after #310 lands.

File: spec/unit/puppet-strings/markdown_spec.rb [hunk 3]
```diff
@@ -2,56 +2,55 @@
...
-    YARD::Parser::SourceParser.parse_string(fixture_content("ruby/resource_type.rb"), :ruby)
-    YARD::Parser::SourceParser.parse_string(fixture_content("ruby/resource_api.rb"), :ruby)
-
-    # task metadata derives the task name from the filename, so we have to parse
-    # directly from the filesystem to correctly pick up the name
-    YARD::Parser::SourceParser.parse(File.join(fixture_path, "json/backup.json"))
-  end
+    def parse_shared_content
+      # Populate the YARD registry with both Puppet and Ruby source
+      YARD::Parser::SourceParser.parse_string(fixture_content("puppet/class.pp"), :puppet)
```
Comment (ekohl) [inline spec/unit/puppet-strings/markdown_spec.rb:20]: Isn't it tempting to rewrite this to a loop?

File: lib/puppet-strings/markdown/data_types.rb [hunk 4]
```diff
@@ -3,41 +3,19 @@
 require_relative 'data_type'
 
 module PuppetStrings::Markdown
+  # Adapter to get information about data types
   module DataTypes
```
Comment (ekohl) [inline lib/puppet-strings/markdown/data_types.rb:7]: I was really wondering if this class shouldn't be killed off instead. Just merge it into `DataType`. Possibly better for a future PR.

**Review by ekohl — APPROVED** [review]
So tempting to remove even more code.

File: lib/puppet-strings/markdown.rb [hunk 5]
```diff
@@ -41,4 +70,17 @@ def self.render(path = nil)
       YARD::Logger.instance.debug "Wrote markdown to #{path}"
     end
   end
+
+  # Helper function to load an ERB template.
+  #
+  # @param [String] path The full path to the template file.
+  # @return [ERB] Template
+  def self.erb(path)
+    if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.6.0')
```
[the remark this answers is not in the document]
Comment (danielparks) [reply lib/puppet-strings/markdown.rb:79]: Hm… I thought about it, but I’m inclined not to since it ends up duplicating the doc comment. Or maybe I should just put the docs above the outer `if` if I do that? Pretty sure the performance difference will be negligible.

File: spec/unit/puppet-strings/markdown_spec.rb (hunk 3 above)
Comment (danielparks) [reply spec/unit/puppet-strings/markdown_spec.rb]: Ha! This has already started snowballing a bit… I keep coming back and finding little tweaks, and then those bring more things in scope that I could tweak…

File: lib/puppet-strings/markdown/data_types.rb (hunk 4 above)
Comment (danielparks) [reply lib/puppet-strings/markdown/data_types.rb]: Ah ha! I’d been thinking about different ways to clean this up, but for some reason that didn’t occur to me. It makes it a lot cleaner.

**Review by bastelfreak — APPROVED** [review]

**Review by chelnak — APPROVED** [review]

File: lib/puppet-strings/markdown/base.rb [hunk 6]
```diff
@@ -50,6 +50,33 @@ module PuppetStrings::Markdown
   #  ") inherits foo::bar {\n" +
   #  "}"}
   class Base
+    # Set or return the name of the group
+    #
+    # @param [Optional[String]] Name of the group to set
+    # @return [String] Name of the group
+    def self.group_name(name = nil)
```
Comment (ekohl) [inline lib/puppet-strings/markdown/base.rb:57]: Thoughts about splitting the functionality in `self.group_name()` and `self.group_name=(name)`? Would that work?

Same comment applies to `yard_types`.

File: lib/puppet-strings/markdown/base.rb [hunk 7]
```diff
@@ -50,6 +50,33 @@ module PuppetStrings::Markdown
...
+      @yard_types
+    end
+
+    # @return [Array] list of items
+    def self.items
+      yard_types
+        .flat_map { |type| YARD::Registry.all(type) }
+        .sort_by(&:name)
+        .map(&:to_hash)
+        .map { |i| new(i) }
```
Comment (ekohl) [inline lib/puppet-strings/markdown/base.rb:77]: You can combine both maps
```suggestion
        .map { |i| new(i.to_hash) }
```

File: lib/puppet-strings/markdown/base.rb [hunk 8]
```diff
@@ -50,6 +50,33 @@ module PuppetStrings::Markdown
...
+    # @return [Array[Symbol]] Array of symbols registered with YARD
+    def self.yard_types(types = nil)
+      @yard_types = types if types
+      @yard_types
+    end
+
+    # @return [Array] list of items
+    def self.items
+      yard_types
+        .flat_map { |type| YARD::Registry.all(type) }
```
Comment (ekohl) [inline lib/puppet-strings/markdown/base.rb:74]: TIL about https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-flat_map. Neat!

File: lib/puppet-strings/markdown/base.rb (hunk 6 above)
Comment (danielparks) [reply lib/puppet-strings/markdown/base.rb:57]: I actually tried doing

```ruby
class << self
  attr_accessor :group_name
  attr_accessor :yard_types
end
```

but I ran into a few problems. One was that it interpreted `yard_types = [...]` as trying to set a local variable instead of calling a class method, but maybe I could have called `self.yard_types = [...]`.

I wanted it to look more like functions like `attr_accessor` since I figured that would feel more natural, but it did seem inconsistent with the way Ruby does things elsewhere.

I don’t have a real good sense for the Ruby Way anymore… been writing other languages for too long.

File: lib/puppet-strings/markdown/base.rb (hunk 7 above)
Comment (danielparks) [reply lib/puppet-strings/markdown/base.rb:77]: 🤦

File: lib/puppet-strings/markdown/base.rb (hunk 6 above)
Comment (ekohl) [reply lib/puppet-strings/markdown/base.rb:57]: Fair enough. I suppose this is good enough.
